### ISSUE TITLE
C++: badly parsed expressions bring down the whole item

### ISSUE THREAD

**Issue by other-mickk** [sphinx-doc/sphinx#4114]
Running a fresh copy of master (c57cd9cb064e63b8303697a31c97679dd8b09d9a) with Python 3.6. Some synthesised examples which can be run on a stock project:

```rst
.. cpp:var:: int one = 1

   This one is fine.

.. cpp:var:: int zero = 0

   (1)

.. cpp:var:: int type_size = sizeof...(Args)

   (2)

References: :cpp:any:`one`, :cpp:any:`zero`, :cpp:any:`type_size`.
```

The warning for (1):

```text
WARNING: /home/mickk/DEV/sphinx/demo/source/index.rst:5: (WARNING/2) Error in postfix expression, expected primary expression or type.
If primary expression:
  Invalid definition: Expected identifier in nested name. [error at 11]
    int zero = 0
    -----------^
If type:
  Invalid definition: Expected identifier in nested name. [error at 11]
    int zero = 0
    -----------^
```

For (2):

```text
WARNING: /home/mickk/DEV/sphinx/demo/source/index.rst:9: (WARNING/2) Error in postfix expression, expected primary expression or type.
If primary expression:
  Invalid definition: Expected identifier in nested name, got keyword: sizeof [error at 22]
    int type_size = sizeof...(Args)
    ----------------------^
If type:
  Invalid definition: Expected identifier in nested name, got keyword: sizeof [error at 22]
    int type_size = sizeof...(Args)
```

I understand that parsing of expressions is being worked on. My concerns are not so much about the limitations of the parser themselves (when it comes to C++ there'll always be a too-clever-by-half, impossibly hard to parse expression to construct,  right? ;), but more about the fact that currently failing to parse voids the whole item: none of `zero` or `type_size` are rendered prettily, and they can’t be linked to either.

I produced the examples from some actual documentation of mine. Compared to the current behaviour of leaving the expressions 'opaque' while still styling & linking to the items, it would mean the changes to parsing would be breaking.

**Comment by jakobandersen** [comment]
Thanks for the report. The parsing error with (1) is strange, while (2) is not yet supported. It's a good point that expression parsing shouldn't ruin the typesetting of the rest of declarations. I'll make it fall back tto the old parser, with an additional warning about it.

**Comment by jakobandersen** [comment]
The cases should now work, and the fallback parser is back.

**Comment by other-mickk** [comment]
Could this have introduced the following bug (which should be a regression):

```rst
.. cpp:concept:: template<typename... Pack> Demo = sizeof...(Pack)
```

The tail-end of the traceback:

```text
  File "/home/mickk/DEV/PyEnvs/sphinx/lib/python3.6/site-packages/Sphinx-1.7.dev20171127-py3.6.egg/sphinx/domains/cpp.py", line 871, in describe_signature
    self.exprs[0].describe_signature(signode, mode, env, symbol)
  [Previous line repeated 7 more times]
  File "/home/mickk/DEV/PyEnvs/sphinx/lib/python3.6/site-packages/Sphinx-1.7.dev20171127-py3.6.egg/sphinx/domains/cpp.py", line 963, in describe_signature
    self.identifier.describe_signature(signode, mode, env, symbol, "")
  File "/home/mickk/DEV/PyEnvs/sphinx/lib/python3.6/site-packages/Sphinx-1.7.dev20171127-py3.6.egg/sphinx/domains/cpp.py", line 1202, in describe_signature
    targetText = prefix + self.identifier
TypeError: unsupported operand type(s) for +: 'Symbol' and 'str'
```

I can open a new issue if needed.

**Comment by jakobandersen** [comment]
Thanks, this was indeed a bug in the new code. It should be fixed now.

**Comment by other-mickk** [comment]
@jakobandersen I have another edge case:

```rst
.. cpp:type:: std::ostream

.. cpp:concept:: template<typename Type> Trivial

    The trivial concept.

.. cpp:type:: template<typename Unconstrained = std::ostream&> unconstrained_param

    This one is fine.

.. cpp:type:: template<Trivial Constrained = std::ostream> constrained_param

    This one is fine.

.. cpp:type:: template<Trivial Constrained = std::ostream&> constrained_param_with_ref

    This one is not.

.. cpp:type:: template<int = qualified::id&> nonsensical

    A nonsensical but standalone repro.
```

The last two items fail with:

```text
/home/mickk/DEV/sphinx/demo/source/index.rst:15: WARNING: Error in template parameter list.
Invalid definition: Expected "=", ",", or ">". [error at 43]
  template<Trivial Constrained = std::ostream&> constrained_param_with_ref
  -------------------------------------------^
/home/mickk/DEV/sphinx/demo/source/index.rst:19: WARNING: Error in template parameter list.
Invalid definition: Expected "=", ",", or ">". [error at 28]
  template<int = qualified::id&> nonsensical
  ----------------------------^
```

This one is really specific (aka parsing a type vs parsing an expression) so I don't mind if it's not really supported (`nonsensical` in particular does *not* need to work). However trying it on master right now the items don't get rendered and refs don't link, which is a regression.

**Comment by jakobandersen** [comment]
Thanks, this was a general problem with defaulted constrained type parameters, and should be fixed in c7a145cc18.
If few notes:
- Before it parsed an expression, but now tries first an expression and then a type.
- For this to work the expression fallback parser is now disabled in this specific case.
- In general Sphinx can not know if an expression or a type is the correct choice, as it doesn't check semantics, and the type/concept (e.g., ``Trivial``) may not be documented at all.
- This means that there are no errors in ``nonsensical``.
